### PR TITLE
feat: deleted the pods that are not unmanaged by Cilium

### LINKED ISSUE #21594: Unmanaged pods with preemptible nodes on GKE
### Is there an existing issue for this?

- [X] I have searched the existing issues

### What happened?

Our organisation is using GKE with [preemptible](https://cloud.google.com/compute/docs/instances/preemptible) node pools. The preemptible VMs used in those node pools aren't _always available_ so we see some VMs being stopped and recreated from time to time. We use these VMs to reduce cost and with proper taints, tolerations & node affinity configuration we're able to run non-sensitive workloads on these node pools while running other more sensitive workloads on proper VMs. 

When using Cilium with preemptible VMs in GKE, we've noticed that when a preemptible VM linked to a node is deleted, the underlying compute instance in GCP gets recreated but the k8s node _remains the same_. Because the k8s node isn’t recreated the node pool taint doesn’t get reapplied and we end up in situations where pods are potentially unmanaged if cilium isn't up and ready when other pods get created. Ultimately in these situations we then see legitimate traffic being blocked.

I wonder if somehow the `cilium-operator` could help in this situation: from what I can see the operator is already responsible for watching cilium-agent pods and removing the node taint once cilium is ready. Would it somehow be possible to taint back a node when the operator sees the cilium-agent not running/modified or deleted on a node?

### Cilium Version

Cilium v1.12.0

### Kernel Version

COS-5.10.133

### Kubernetes Version

v1.22.12-gke.1200

### Sysdump

_No response_

### Relevant log output

_No response_

### Anything else?

I did pop a message in the Cilium kubernetes slack channel too: https://cilium.slack.com/archives/C53TG4J4R/p1664808717519899  

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by lvyanru8200**
Set operator to remove the label of a pod that existed before the node taint

1. Delete the specified label pod according to the parameter --pod-restart-selector, default value is k8s-app=kube-dns
2. --pod-restart-selector="" Remove all pods

Fixes: #21594
Signed-off-by: yanru.lv <yanru.lv@daocloud.io>

Please ensure your pull request adheres to the following guidelines:

- [ ] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [ ] All code is covered by unit and/or runtime tests where feasible.
- [ ] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [ ] If your commit description contains a `Fixes: <commit-id>` tag, then
      please add the commit author[s] as reviewer[s] to this issue.
- [ ] All commits are signed off. See the section [Developer’s Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [ ] Provide a title or release-note blurb suitable for the release notes.
- [ ] Thanks for contributing!

Fixes: #21594

```release-note
<!-- Enter the release note text here if needed or remove this section! -->
```

### CODE REVIEW

File: operator/cmd/flags.go [hunk 1]
```diff
@@ -324,6 +324,9 @@ func init() {
 	flags.Bool(operatorOption.EnableK8s, true, `Enable operation of Kubernetes-related services/controllers when using Cilium with Kubernetes`)
 	option.BindEnv(Vp, operatorOption.EnableK8s)
 
+	flags.String(operatorOption.PodRestartSelector, "k8s-app=kube-dns", "Set operator to remove the label of a pod that existed before the node taint")
```
Comment (joestringer) [inline operator/cmd/flags.go]:
```suggestion
	flags.String(operatorOption.PodRestartSelector, "k8s-app=kube-dns", "cilium-operator will delete/restart any pods with these labels if the pod is not managed by Cilium. If this option is empty, then all pods may be restarted.")
```

File: operator/watchers/pod.go [hunk 2]
```diff
@@ -4,6 +4,7 @@
 package watchers
 
 import (
+	operatorOption "github.com/cilium/cilium/operator/option"
```
Comment (joestringer) [inline operator/watchers/pod.go]: Please move this line down next to the other `github.com/cilium/cilium` imports.

**Review by joestringer — CHANGES_REQUESTED** [review]
Couple of minor comments, please change these.

Also, just to let you know: you do not need to create new PRs like this each time you make changes. It should be possible to `git push --force` to override the previous PR's content on the branch. This will refresh the GitHub page and highlight the latest version of your code.

**Comment by lvyanru8200** [comment]
cc @joestringer

**Review by aanm — CHANGES_REQUESTED** [review]
Just one small change, LGTM.

**Review by aanm — APPROVED** [review]
Oh I see that Joe already requested for the same change that I did. In that case I'm approving as long Joe's review is addressed.

**Review by nebril — APPROVED** [review]

**Comment by lvyanru8200** [comment]
@joestringer ping

**Comment by joestringer** [comment]
Travis failed, please resolve the issue: https://app.travis-ci.com/github/cilium/cilium/jobs/590341246 .

**Comment by lvyanru8200** [comment]
> Travis failed, please resolve the issue: https://app.travis-ci.com/github/cilium/cilium/jobs/590341246 .

How to re-trigger the Travis CI process please，I checked the logs and found no more obvious errors

**Comment by christarazi** [comment]
@lvyanru8200 What's missing is that the cmdref documention has not been updated. Could you run `make -j $(nproc) && make -j $(nproc) -C Documentation/ update-cmdref` and then amend the current commit, and force push?

**Comment by lvyanru8200** [comment]
@christarazi thanks, @joestringer  cc

**Comment by lvyanru8200** [comment]
@christarazi https://github.com/cilium/cilium/actions/runs/3674055290/jobs/6211823310，Is this error a go version issue? Doesn't feel like my fault

**Comment by christarazi** [comment]
No, I think this has been failing across other PRs as well.

**Comment by lvyanru8200** [comment]
@joestringer ping

**Review by joestringer — APPROVED** [review]

**Comment by christarazi** [comment]
@lvyanru8200 Please rebase your PR instead of introducing a merge commit, as we don't accept those in the tree.

**Pull request closed without merging**